pkg/tests/api/alerting/api_alertmanager_test.go

```diff
@@ -1395,6 +1397,230 @@ func TestAlertRuleCRUD(t *testing.T) {
+						Data: []ngmodels.AlertQuery{
+							{
+								RefID: "A",
+								RelativeTimeRange: ngmodels.RelativeTimeRange{
+									From: ngmodels.Duration(time.Duration(5) * time.Hour),
+									To:   ngmodels.Duration(time.Duration(3) * time.Hour),
+								},
+								DatasourceUID: "-100",
+								Model: json.RawMessage(`{
+												"type": "math",
+												"expression": "2 + 3 < 1"
+												}`),
+							},
+						},
+						NoDataState:  apimodels.NoDataState(ngmodels.Alerting),
+						ExecErrState: apimodels.ExecutionErrorState(ngmodels.AlertingErrState),
+					},
+				},
+			},
+		}
+		buf := bytes.Buffer{}
+		enc := json.NewEncoder(&buf)
+		err = enc.Encode(&rules)
+		require.NoError(t, err)
+
+		u := fmt.Sprintf("http://grafana:password@%s/api/ruler/grafana/api/v1/rules/default", grafanaListedAddr)
+		// nolint:gosec
+		resp, err := http.Post(u, "application/json", &buf)
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			err := resp.Body.Close()
+			require.NoError(t, err)
+		})
+		b, err := ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		assert.Equal(t, resp.StatusCode, 202)
+		require.JSONEq(t, `{"message":"rule group updated successfully"}`, string(b))
+
+		// let's make sure that rule definitions are updated correctly.
+		u = fmt.Sprintf("http://grafana:password@%s/api/ruler/grafana/api/v1/rules/default", grafanaListedAddr)
+		// nolint:gosec
+		resp, err = http.Get(u)
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			err := resp.Body.Close()
+			require.NoError(t, err)
+		})
+		b, err = ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		assert.Equal(t, resp.StatusCode, 202)
+
+		body, m := rulesNamespaceWithoutVariableValues(t, b)
+		returnedUIDs, ok := m["default,arulegroup"]
+		assert.True(t, ok)
+		assert.Equal(t, 1, len(returnedUIDs))
+		assert.Equal(t, ruleUID, returnedUIDs[0])
+		assert.JSONEq(t, `
+			{
+			   "default":[
+			      {
+				 "name":"arulegroup",
+				 "interval":"1m",
+				 "rules":[
+				    {
+				       "expr":"",
+				       "for": "30s",
+				       "grafana_alert":{
+					  "id":1,
+					  "orgId":1,
+					  "title":"AlwaysNormal",
+					  "condition":"A",
+					  "data":[
+					     {
+						"refId":"A",
+						"queryType":"",
+						"relativeTimeRange":{
+						   "from":18000,
+						   "to":10800
+						},
+						"datasourceUid":"-100",
+									"model":{
+						   "expression":"2 + 3 \u003C 1",
+						   "intervalMs":1000,
+						   "maxDataPoints":43200,
+						   "type":"math"
+						}
+					     }
+					  ],
+					  "updated":"2021-02-21T01:10:30Z",
+					  "intervalSeconds":60,
+					  "version":3,
+					  "uid":"uid",
+					  "namespace_uid":"nsuid",
+					  "namespace_id":1,
+					  "rule_group":"arulegroup",
+					  "no_data_state":"Alerting",
+					  "exec_err_state":"Alerting"
+				       }
+				    }
+				 ]
+			      }
+			   ]
+			}`, body)
+	}
+
+	// update the rule; keep title, condition, interval, for, queries and expressions if not provided
```

Comment:
Is this case meant to actually update anything? IE Are `NoDataState` + `ExecErrState` meant to be the same as before?

Reply:
good question; I don't think that the UI will ever send empty values for those states but I have change the behaviour and keep them as well.